src/java/org/apache/cassandra/schema/PartitionDenylist.java

```diff
@@ -0,0 +1,462 @@
+                                       {
+                                           final DenylistEntry newEntry = getDenylistForTable(tid);
+                                           if (newEntry != null)
+                                               return newEntry;
+                                           if (oldValue != null)
+                                               return oldValue;
+                                           return new DenylistEntry();
+                                       }
+                                   });
+                                   executor.execute(task);
+                                   return task;
+                               }
+                           });
+    }
+
+    /**
+     * We need to fully rebuild a new cache to accommodate deleting items from the denylist and potentially shrinking
+     * the max allowable size in the list. We do not serve queries out of this denylist until it is populated
+     * so as not to introduce a window of having a partially filled cache allow denylisted entries.
+     */
+    public boolean load()
+    {
+        final long start = System.currentTimeMillis();
+
+        final Map<TableId, DenylistEntry> allDenylists = getDenylistForAllTables();
+        if (allDenylists == null)
+            return false;
+
+        // On initial load we have the slight overhead of GC'ing our initial empty cache
+        LoadingCache<TableId, DenylistEntry> newDenylist = buildEmptyCache();
+        newDenylist.putAll(allDenylists);
+
+        synchronized (this)
+        {
+            loadSuccesses++;
+        }
+        denylist = newDenylist;
+        logger.info("Loaded partition denylist cache in {}ms", System.currentTimeMillis() - start);
+        return true;
+    }
+
+    /**
+     * We expect the caller to confirm that we are working with a valid keyspace and table. Further, we expect the usage
+     * pattern of this to be one-off key by key, not in a bulk process, so we relead the entire table's deny list entry
+     * on an addition or removal.
+     */
+    public boolean addKeyToDenylist(final String keyspace, final String table, final ByteBuffer key)
+    {
+        if (!isPermitted(keyspace))
+            return false;
+
+        final byte[] keyBytes = new byte[key.remaining()];
+        key.slice().get(keyBytes);
+
+        final String insert = String.format("INSERT INTO system_distributed.partition_denylist (ks_name, cf_name, key) VALUES ('%s', '%s', 0x%s)",
+                                            keyspace, table, Hex.bytesToHex(keyBytes));
+
+        try
+        {
+            process(insert, DatabaseDescriptor.getDenylistConsistencyLevel());
+
+            // On insert, we refresh the entire table's deny list data.
+            TableId denyTableId = getTableId(keyspace, table);
+            denylist.put(denyTableId, getDenylistForTable(getTableId(keyspace, table)));
```

Comment:
At this moment we update both in-memory cache and and cql table. And one node starts having different state compared to others. That scares me a bit (as I've already stated in another comment regarding syncing data between caches). Should we just update CQL and ensure data is simply reloaded across the cluster via JMX? It would make more sense from operational standpoint I guess. WDYT?

Reply:
So the scenario causing the concern is:
1. A denylist entry is added for PK N. Node A and B receive it, C does not due to unavailability. Quorum is achieved on the denylist change.
2. A query comes in for PK N to Node B
3. ???

This would depend on the CL of the query on PK N right? If you ask for CL.ONE, you're definitely at risk of getting data that's not deny listed correctly for sure. If you query at any other CL, you should get rejection from Node A and C as they received the denylist update and properly reject the query.

Our options to tighten this up would be to force CL.ALL for denylist modifications (which we effectively do right now in our codepaths for initialLoad relying on sufficientLiveNodes()...) which introduces availability issues to using the feature, or to for a higher CL for mutations to Keyspaces that have deny-listed keys which isn't really an option.

Not updating the data in memory once the denylist CQL mutation succeeds broadens the window of time between the initial expression of intent of an operator (deny access to this key!) data being in CQL and data loading in your denylist cache. While we get the "RF-wide atomicity" property of calling `sufficientLiveNodesForSelectStar` on the cache reload JMX call, it's at the cost of forcing the operator to take another step when they denylist a key as well as not allowing cache refresh unless all nodes in the replica set for the cache are available.

Hm... _I think we do have a problem though_. There's no mechanism for refreshing the in-memory cache based on post-hoc adjustments to the underlying system_distributed table (streaming, repair, hints). While we could hook into `Keyspaces.applyInternal `similarly to how we handle view mutations w/`requiresViewUpdate` and then push denylist changes through the PartitionDenylist mechanisms... I'm not sure we should.

This puts us squarely in the "denylist cache refresh requires all nodes to be up in order for you to update the denied keys in memory" no matter how we slice it, so maybe the initial implementation of just having it be a two step process for operators (1. adjust CQL/JMX, 2. reload your cache) is the right way after all.

Let me think on this a bit more. Interesting problem.

Reply:
Yeah, my point was more about the second part related to sync of denylist cache and CQL table. Also it can be even more complicated if we think about failover scenarios.

Looking forward to hearing your thoughts on the final design.